File: hw/pastor/u3/hw29Singleton/JAVASCRIPT/Config/AlarmConfig.js
import { MongoClient, ObjectId } from "mongodb";

export class AlarmConfig {
    static instance = null;

    constructor() {
        if (AlarmConfig.instance) {
            throw new Error("Use getInstance()");
        }
    }

    static async getInstance() {
        if (!AlarmConfig.instance) {
            const config = new AlarmConfig();
            await config.#init();
            AlarmConfig.instance = config;
        }
        return AlarmConfig.instance;
    }

    async #init() {
        this.client = new MongoClient("mongodb+srv://Mathews:Mathews2007@cluster0.6l9ibfh.mongodb.net/?appName=Cluster0");
        await this.client.connect();

        this.db = this.client.db("SingletonDB");
        this.collection = this.db.collection("sales_configuration");

        const doc = await this.collection.findOne();

        if (!doc) {
            const result = await this.collection.insertOne({ minimumStock: 10 });
            this.id = result.insertedId;
            this.minimumStock = 10;
        } else {
            this.id = doc._id;
            this.minimumStock = doc.minimumStock;
        }
    }

    getMinimumStock() {
        return this.minimumStock;
    }

    async updateMinimumStock(newStock) {
        this.minimumStock = newStock;
        await this.collection.updateOne(
            { _id: this.id },
            { $set: { minimumStock: newStock } }
        );
    }
}

File: hw/pastor/u3/hw29Singleton/PYTHON/Config/AlarmConfig.py
from pymongo import MongoClient

class AlarmConfig:
    __instance = None

    def __init__(self):
        if AlarmConfig.__instance is not None:
            raise Exception("Singleton already exists")

        self.client = MongoClient("mongodb+srv://Mathews:Mathews2007@cluster0.6l9ibfh.mongodb.net/?appName=Cluster0")
        self.db = self.client["SingletonDB"]
        self.collection = self.db["sales_configuration"]

        config = self.collection.find_one({"_id": "alarm_config"})

        if config is None:
            self.minimumStock = 10
            self.collection.insert_one({
                "_id": "alarm_config",
                "minimumStock": self.minimumStock
            })
        else:
            self.minimumStock = config["minimumStock"]

        AlarmConfig.__instance = self

    @staticmethod
    def getInstance():
        if AlarmConfig.__instance is None:
            AlarmConfig()
        return AlarmConfig.__instance

    def getMinimumStock(self):
        return self.minimumStock

    def updateMinimumStock(self, newStock: int):
        self.minimumStock = newStock
        self.collection.update_one(
            {"_id": "alarm_config"},
            {"$set": {"minimumStock": newStock}}
        )
